legacy/reqcounter/reqcounter.go

```diff
@@ -24,28 +24,71 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
-// RequestCounter records the number of HTTP requests to GCR.
-type RequestCounter struct {
-	mutex    sync.Mutex // Lock to prevent race-conditions with concurrent processes.
-	requests uint64     // Number of HTTP requests since recording started.
-	since    time.Time  // When the current counter began recording requests.
+// requestCounter records the number of HTTP requests to GCR.
+// TODO: @tylerferrara in the future, add a field 'persistent bool' to determine if
+// the counter should ever reset.
+type requestCounter struct {
+	mutex    sync.Mutex    // Lock to prevent race-conditions with concurrent processes.
+	requests uint64        // Number of HTTP requests since recording started.
+	since    time.Time     // When the current counter began recording requests.
+	interval time.Duration // The duration of time between each log.
 }
 
 // log records the number of HTTP requests found and resets the counter.
-func (rc *RequestCounter) log() {
+func (rc *requestCounter) log() {
 	// Hold onto the lock when reading & writing the counter.
 	rc.mutex.Lock()
 	defer rc.mutex.Unlock()
 
 	// Log the number of requests within this measurement window.
-	msg := fmt.Sprintf("Since %s there have been %d requests to GCR.", counter.since.Format("2006-01-02 15:04:05"), rc.requests)
+	msg := fmt.Sprintf("Since %s there have been %d requests to GCR.", rc.since.Format("2006-01-02 15:04:05"), rc.requests)
 	logrus.Debug(msg)
 
 	// Reset the counter.
+	rc.reset()
+}
+
+// reset clears the request counter and stamps the current time of reset.
+// TODO: @tylerferrara in the future, use the request counter field 'persistent' to check
+// if the counter should be reset.
+func (rc *requestCounter) reset() {
 	rc.requests = 0
 	rc.since = time.Now()
 }
 
+// watch continuously logs the counter at the specified intervals.
+func (rc *requestCounter) watch() {
+	for {
+		time.Sleep(rc.interval)
+		rc.log()
+	}
+}
+
+// networkMonitor is the primary means of monitoring network traffic between CIP and GCR.
+type networkMonitor struct {
+	counters []*requestCounter
+	mutex    sync.Mutex
+}
+
+// increment adds 1 to each request counter, signifying a new request has been made to GCR.
+func (nm *networkMonitor) increment() {
+	// Hold onto the lock when writing to each counter.
+	nm.mutex.Lock()
+	defer nm.mutex.Unlock()
+
+	// Increment each counter.
+	for _, rc := range nm.counters {
+		rc.requests++
```

Comment:
consider using https://gobyexample.com/atomic-counters instead 

also adding to the above comment, let's move the incrementing logic to the individual requestCounters to avoid directly trying to modify struct variables in a multi-threaded environment.

Reply:
In this case, adding `atomic.AddUint64(&rc.requests, 1)` will not get rid of our mutex. Since we require no modification of the counter within the `log()` function, both `increment()` and `log()` reach for the mutex when attempting any read or write to the counter.

Thank being said, I can most certainly add this, though it won't provide any benefits. Is this still worth pursuing?

Reply:
I think it's fine to use it anyway (which I see in the current version). It would avoid needing a mutex in increment() atleast. 

And also if flush weren't accessing multiple variables, we could've gotten away with atomic.LoadInt and avoid mutexes altogether.

Reply:
I can use the atomic.AddUnit64, but again, I still must keep the mutex in increment(). This is because we could still increment the request when flush holds onto the mutex. In order for race conditions to be avoided, all consumers must be reliant on first grabbing the mutex.